pulsar-common/src/main/java/org/apache/pulsar/common/util/FutureUtil.java

```diff
@@ -55,6 +57,36 @@ public static CompletableFuture<Object> waitForAny(Collection<? extends Completa
         return CompletableFuture.anyOf(futures.toArray(new CompletableFuture[0]));
     }
 
+    public static CompletableFuture<Optional<Object>> waitForAny(Collection<? extends CompletableFuture<?>> futures,
+                                                       Predicate<Object> tester) {
+        return waitForAny(futures).thenCompose(v -> {
+            if (tester.test(v)) {
+                return CompletableFuture.completedFuture(Optional.of(v));
+            }
+            Collection<CompletableFuture<?>> doneFutures = futures.stream()
+                    .filter(f -> f.isDone())
+                    .collect(Collectors.toList());
+            futures.removeAll(doneFutures);
+            Optional<?> value = doneFutures.stream()
+                    .filter(f -> !f.isCompletedExceptionally())
+                    .map(CompletableFuture::join)
```

Comment:
Can we avoid the `join`? async & sync operation.

Reply:
This is for `doneFutures`, all the futures are completed.

Reply:
Thanks for your reply.